src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -674,3 +682,430 @@ def _standard_metadata(self, name: str) -> V1ObjectMeta:
+    ) -> None:
+        obj_name = f"{username}-fs"
+        crd_group = "gafaelfawr.lsst.io"
+        crd_version = "v1alpha1"
+        plural = "gafaelfawringresses"
+        try:
+            await self.custom_api.create_namespaced_custom_object(
+                body=spec,
+                group=crd_group,
+                version=crd_version,
+                namespace=namespace,
+                plural=plural,
+            )
+        except ApiException as e:
+            if e.status == 409:
+                # It already exists.  Delete and recreate it
+                self._logger.warning(
+                    (
+                        "Fileserver gafaelfawringress exists. "
+                        + "Deleting and recreating."
+                    ),
+                    name=obj_name,
+                    namespace=namespace,
+                )
+                await self._delete_fileserver_gafaelfawringress(
+                    username, namespace
+                )
+                await self._wait_for_fileserver_object_deletion(
+                    obj_name=obj_name,
+                    namespace=namespace,
+                    kind="gafaelfawringress",
+                )
+                await self.custom_api.create_namespaced_custom_object(
+                    crd_group, crd_version, namespace, plural, spec
+                )
+            raise KubernetesError.from_exception(
+                "Error creating gafaelfawringress",
+                e,
+                namespace=namespace,
+                name=obj_name,
+            ) from e
+
+    async def _delete_fileserver_gafaelfawringress(
+        self, username: str, namespace: str
+    ) -> None:
+        obj_name = f"{username}-fs"
+        crd_group = "gafaelfawr.lsst.io"
+        crd_version = "v1alpha1"
+        plural = "gafaelfawringresses"
+        try:
+            await self.custom_api.delete_namespaced_custom_object(
+                crd_group, crd_version, namespace, plural, obj_name
+            )
+        except ApiException as e:
+            if e.status == 404:
+                return
+            raise KubernetesError.from_exception(
+                "Error deleting gafaelfawringress",
+                e,
+                namespace=namespace,
+                name=obj_name,
+            ) from e
+
+    async def get_observed_fileserver_state(
+        self, namespace: str
+    ) -> dict[str, bool]:
+        """Reconstruct the fileserver user map with what we can determine
+        from the Kubernetes cluster.
+
+        Objects with the name <username>-fs are presumed to be fileserver
+        objects, where <username> can be assumed to be the name of the
+        owning user.
+
+        It returns a dict mapping strings to the value True, indicating those
+        users who currently have fileservers.
+        """
+        observed_state: dict[str, bool] = {}
+        # Create fileserver namespace if necessary.
```

Comment:
Stray comment; there's no code here to create it.

Reply:
Removed